### PR TITLE
Manifold-dependent allocation

### PR DESCRIPTION
**Pull request by mateuszbaran**
Another point from https://github.com/JuliaManifolds/Manifolds.jl/issues/497 .

### CODE REVIEW

**Comment by mateuszbaran** [comment]
Ah, `StaticArraysCore` is not available for Julia 1.0. So now I'm not sure where to put
```julia
function allocate(::PowerManifoldNestedReplacing, x::AbstractArray{<:SArray})
    return similar(x)
end
```

**Comment by mateuszbaran** [comment]
Some parts of nested work with plain base but we extend it in Manifolds.jl anyway. I think I'll do just that.

File: src/ManifoldsBase.jl [hunk 1]
```diff
@@ -36,26 +36,46 @@ include("projections.jl")
...
 allocation and is forwarded to the function `similar`.
+
+It's behavior can be overriden by a specific manifold, for example power manifold with
+nested replacing representation can decide that `allocate` for `Array{<:SArray}` returns
+another `Array{<:SArray}` instead of `Array{<:MArray}`, as would be done by default.
 """
 allocate(a, args...)
 allocate(a) = similar(a)
-allocate(a, dims::Integer...) = similar(a, dims...)
+allocate(a, dim1::Integer, dims::Integer...) = similar(a, dim1, dims...)
```
Comment (kellertuer) [inline src/ManifoldsBase.jl:59]: Why the `dim1`? that is not documented above.

Comment (mateuszbaran) [reply src/ManifoldsBase.jl:59]: It's there to guarantee that this specialization only applies to calls with at least one integer, it's essentially an implementation detail. I'm not even sure if entirely necessary but at one point it showed up in an ambiguity list.

Comment (kellertuer) [reply src/ManifoldsBase.jl:59]: But aren't the integers anyways collected in `dims...`?

Comment (mateuszbaran) [reply src/ManifoldsBase.jl:59]: Yes, though the difference is that `allocate(a, dims::Integer...)` can be called without any integers (`allocate([1])`) while `allocate(a, dim1::Integer, dims::Integer...)` ensures that there is at least one integer.

Comment (kellertuer) [reply src/ManifoldsBase.jl:59]: ok, if you feel that thats necessary.

**Review by kellertuer — APPROVED** [review]

**Pull request merged**